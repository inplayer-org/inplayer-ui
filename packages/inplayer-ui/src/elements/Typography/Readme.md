Typography is an element which allows you to control the style of your text.
It has a `variant` prop which allows you to choose between the h1-h6 headings.
It inherits the `palette.primary.dark` text color, however this can be overridden by
using styled-components.

```jsx static
import { Typography } from '@inplayer-org/inplayer-ui';

<div>
  <Typography variant="h1">Heading H1</Typography>
  <Typography variant="h2">Heading H2</Typography>
  <Typography variant="h3">Heading H3</Typography>
  <Typography variant="h4">Heading H4</Typography>
  <Typography variant="h5">Heading H5</Typography>
  <Typography variant="h6">Heading H6</Typography>
  <Typography variant="p">Paragraph</Typography>
  <Typography variant="p" description>
    Paragraph as description
  </Typography>
</div>;
```

### Demo

```jsx
<Typography variant="h1">Heading H1</Typography>
<Typography variant="h2">Heading H2</Typography>
<Typography variant="h3">Heading H3</Typography>
<Typography variant="h4">Heading H4</Typography>
<Typography variant="h5">Heading H5</Typography>
<Typography variant="h6">Heading H6</Typography>
<Typography variant="p">This is a paragraph</Typography>
<Typography variant="p" description>This is a paragraph that is used as a description</Typography>
```

### Typegraphy with modifiers

```jsx
<Typography variant="h1" modifiers={['textPrimary']}>Heading H1 with primary modifier</Typography>
<Typography variant="h2" modifiers={['textDanger']}>Heading H1 with danger modifier</Typography>
<Typography variant="h3" modifiers={['textSuccess']}>Heading H1 with success modifier</Typography>
<Typography variant="h4" modifiers={['textWarning']}>Heading H1 with warning modifier</Typography>
```

### Typegraphy with custom color provideed

```jsx
<Typography variant="h1" color="red">Heading H1 with custom red color</Typography>
<Typography variant="h2" color="mint">Heading H1 with custom mint color</Typography>
<Typography variant="h3" color="purple">Heading H1 with custom purple color</Typography>
<Typography variant="h4" color="black">Heading H1 with custom black color</Typography>
```
