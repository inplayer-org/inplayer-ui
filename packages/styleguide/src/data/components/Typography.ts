import { ComponentData } from "../types";

const usage = `
  import { Typography } from '@inplayer-org/inplayer-ui';

  <div>
    <Typography variant="h1">Heading H1</Typography>
    <Typography variant="h2">Heading H2</Typography>
    <Typography variant="h3">Heading H3</Typography>
    <Typography variant="h4">Heading H4</Typography>
    <Typography variant="h5">Heading H5</Typography>
    <Typography variant="h6">Heading H6</Typography>
    <Typography variant="p">Paragraph</Typography>
    <Typography variant="p" description>
      Paragraph as description
    </Typography>
  </div>;
`;

const demoCode = `
<div>
  <Typography variant="h1">Heading H1</Typography>
  <Typography variant="h2">Heading H2</Typography>
  <Typography variant="h3">Heading H3</Typography>
  <Typography variant="h4">Heading H4</Typography>
  <Typography variant="h5">Heading H5</Typography>
  <Typography variant="h6">Heading H6</Typography>
  <Typography variant="p">This is a paragraph</Typography>
  <Typography variant="p" description>This is a paragraph that is used as a description
  </Typography>
</div>
`;

const demoCodeWithModifiers = `
<div>
  <Typography variant="h1" modifiers={['textPrimary']}>
    Heading H1 with primary modifier
  </Typography>
  <Typography variant="h2" modifiers={['textDanger']}>
    Heading H1 with danger modifier
  </Typography>
  <Typography variant="h3" modifiers={['textSuccess']}>
    Heading H1 with success modifier
  </Typography>
  <Typography variant="h4" modifiers={['textWarning']}>
    Heading H1 with warning modifier
  </Typography>
</div>
`;

const demoCodeCustomColors = `
<div>
  <Typography variant="h1" color="red">Heading H1 with custom red color</Typography>
  <Typography variant="h2" color="mint">Heading H1 with custom mint color</Typography>
  <Typography variant="h3" color="purple">Heading H1 with custom purple color</Typography>
  <Typography variant="h4" color="black">Heading H1 with custom black color</Typography>
</div>
`;

const Typography: ComponentData = {
  description: `
    Typography is an element which allows you to control the style of your text.
    It has a variant prop which allows you to choose between the h1-h6 headings. It inherits the palette.primary.dark text color,
    however this can be overridden by using styled-components.`,
  path: `src/elements/Typography/index.js`,
  usage,
  examples: [
    {
      title: 'Demo',
      code: demoCode,
    },
    {
      title: 'Typegraphy with modifiers',
      code: demoCodeWithModifiers,
    },
    {
      title: 'Typegraphy with custom color provideed',
      code: demoCodeCustomColors,
    }
  ], 
  propsAndMethods: [
    {name: 'children', type: 'Node', default: 'Required'},
    {name: 'variant', type: 'enum', default: 'Required', description: 'One of: h1, h2, h3, h4, h5, h6, p'},
    {name: 'className', type: 'string', description: 'A className can be passed down for further styling or extending with CSS-in-JS'},
    {name: 'color', type: 'string'},
    {name: 'description', type: 'boolean', default: 'false'},
    {name: 'modifiers', type: 'Array<String>	', default: '[]'},
    {name: 'style', type: 'Object', default: '[]'},
  ] 
}

export default Typography;