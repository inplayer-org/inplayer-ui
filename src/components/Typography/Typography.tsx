import React, { ReactChild } from 'react';
import styled from 'styled-components';
import { textPrimary, textDanger, textSuccess, textWarning } from 'modifiers';
import { applyStyleModifiers } from 'styled-components-modifiers';
import colors from 'theme/colors';

const typographyModifiers = {
  textPrimary,
  textDanger,
  textSuccess,
  textWarning,
};

export type TypographyVariant = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';

type ParagraphProps = {
  color: string;
  description: boolean;
  modifiers: any;
};

type HeadingProps = {
  color: string;
  variant: keyof {
    h1: string;
    h2: string;
    h3: string;
    h4: string;
    h5: string;
    h6: string;
  };
  modifiers: Array<string>;
};

const Paragraph = styled.p<ParagraphProps>`
  font-size: ${({ theme }) => theme.font.sizes.medium};
  color: ${({ color, theme }) => color || `${theme.palette.text.main}`};
  ${({ description, theme: { palette, font } }) =>
    description &&
    `
     color: ${palette.text.light};
     font-size:${font.sizes.large};
   `}
  ${applyStyleModifiers(typographyModifiers)}
`;

const Heading = styled.h1<HeadingProps>`
  font-family: ${({ theme }) => theme.font.primary};
  font-weight: ${({ theme }) => theme.font.weights.thin};
  color: ${({ color }) => color || colors.blue};
  font-size: ${({ theme, variant }) => theme.font.sizes[variant]};
  ${applyStyleModifiers(typographyModifiers)}
`;

export type Props = {
  /**
   * Variant type
   */
  variant?: TypographyVariant;
  /**
   * When prop is set, the paragraph element has `palette.primary.fontGray` color
   */
  description?: boolean;
  /**
   * Text color modifiers
   */
  modifiers?: Array<string>;
  /**
   * Text color
   */
  color?: string;
  /**
   * External class name
   */
  className?: string;
  children: ReactChild;
};

const Typography = ({
  variant = 'h1',
  children,
  description = false,
  modifiers = [],
  color = '',
  className = '',
}: Props) => {
  if (variant === 'p') {
    return (
      <Paragraph
        className={className}
        description={description}
        modifiers={modifiers}
        color={color}
      >
        {children}
      </Paragraph>
    );
  }
  return (
    <Heading
      className={className}
      as={variant}
      variant={variant}
      modifiers={modifiers}
      color={color}
    >
      {children}
    </Heading>
  );
};

export default Typography;
