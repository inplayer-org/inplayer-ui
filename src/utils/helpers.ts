import moment from 'moment';

export const getMonthOptions = (): Array<any> =>
  moment.months().map((label, value) => ({ value, displayName: label }));

export const getYearOptions = () => {
  const yearOptions: Array<any> = [];
  for (let i = moment().year() - 100; i <= moment().year(); i += 1) {
    yearOptions.push({ value: i.toString(), displayName: i.toString() });
  }

  return yearOptions;
};
