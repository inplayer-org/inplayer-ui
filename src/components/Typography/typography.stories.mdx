import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Typography from './'

<Meta title="Typography" component={<Typography/>} />

# Typography

Typography is an element which allows you to control the style of your text. 
It has a `variant` prop which allows you to choose between the h1-h6 headings. 
It inherits the `palette.primary.dark` text color, however this can be overridden by using styled-components.

```jsx
import { Component } from '@inplayer-org/inplayer-ui';

const Page = () => (
  <Typography />
)
```

<Props of={Typography}/>

### Default Typography
<Preview withSource="open">
  <Story name="default">
    <div>
        <Typography variant="h1">Heading H1</Typography>
        <Typography variant="h2">Heading H2</Typography>
        <Typography variant="h3">Heading H3</Typography>
        <Typography variant="h4">Heading H4</Typography>
        <Typography variant="h5">Heading H5</Typography>
        <Typography variant="h6">Heading H6</Typography>
        <Typography variant="p">Paragraph</Typography>
        <Typography variant="p" description>
            Paragraph as description
        </Typography>
    </div>
  </Story>
</Preview>

### Typography with modifiers
<Preview withSource="open">
  <Story name="modifiers">
    <div>
        <Typography variant="h1" modifiers={['textPrimary']}>Heading H1 with primary modifier</Typography>
        <Typography variant="h2" modifiers={['textDanger']}>Heading H1 with danger modifier</Typography>
        <Typography variant="h3" modifiers={['textSuccess']}>Heading H1 with success modifier</Typography>
        <Typography variant="h4" modifiers={['textWarning']}>Heading H1 with warning modifier</Typography>
    </div>
  </Story>
</Preview>

### Typography with custom color
<Preview withSource="open">
  <Story name="color">
    <div>
        <Typography variant="h1" color="red">Heading H1 with custom red color</Typography>
        <Typography variant="h2" color="green">Heading H1 with custom mint color</Typography>
        <Typography variant="h3" color="purple">Heading H1 with custom purple color</Typography>
        <Typography variant="h4" color="black">Heading H1 with custom black color</Typography>
    </div>
  </Story>
</Preview>

