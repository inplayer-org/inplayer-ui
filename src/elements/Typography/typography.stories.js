import React from 'react';
import styled from 'styled-components';
import { Colors } from 'src/index';
import Typography from './index';

export default {
  component: Typography,
  title: 'Typography',
};

const Container = styled.div`
  margin: auto;
  width: 90%;
`;
const BoxContainer = styled.div`
  border: 1px solid ${Colors.darkGray};
  width: 90%;
  padding-left: 40px;
`;
const HeadlineTypography = styled(Typography)`
  color: ${Colors.black};
`;
export const Demo = props => (
  <Container>
    <HeadlineTypography variant="h3">Typography</HeadlineTypography>
    <Typography variant="p" description>
      Typography is an element which allows you to control the style of your text. It has a
      `variant` prop which allows you to choose between the h1-h6 headings. It inherits the
      `palette.primary.dark` text color, however this can be overridden by using styled-components.
    </Typography>
    <HeadlineTypography variant="h4">Demo</HeadlineTypography>
    <BoxContainer>
      <Typography variant="h1">Heading H1</Typography>
      <Typography variant="h2">Heading H2</Typography>
      <Typography variant="h3">Heading H3</Typography>
      <Typography variant="h4">Heading H4</Typography>
      <Typography variant="h5">Heading H5</Typography>
      <Typography variant="h6">Heading H6</Typography>
      <Typography variant="p">Paragraph</Typography>
      <Typography variant="p" description>
        Paragraph as description
      </Typography>
    </BoxContainer>
  </Container>
);
export const TypographyWithModifires = props => (
  <Container>
    <HeadlineTypography variant="h4">Typography With Modifires</HeadlineTypography>
    <BoxContainer>
      <Typography variant="h1" modifiers={['textPrimary']}>
        Heading H1 with primary modifier
      </Typography>
      <Typography variant="h2" modifiers={['textDanger']}>
        Heading H1 with danger modifier
      </Typography>
      <Typography variant="h3" modifiers={['textSuccess']}>
        Heading H1 with success modifier
      </Typography>
      <Typography variant="h4" modifiers={['textWarning']}>
        Heading H1 with warning modifier
      </Typography>
    </BoxContainer>
  </Container>
);
export const TypographyWithCustomColorProvideed = props => (
  <Container>
    <HeadlineTypography variant="h4">Typography With Custom Color Provideed</HeadlineTypography>
    <BoxContainer>
      <Typography variant="h1" color="red">
        Heading H1 with custom red color
      </Typography>
      <Typography variant="h2" color="mint">
        Heading H1 with custom mint color
      </Typography>
      <Typography variant="h3" color="purple">
        Heading H1 with custom purple color
      </Typography>
      <Typography variant="h4" color="black">
        Heading H1 with custom black color
      </Typography>
    </BoxContainer>
  </Container>
);
