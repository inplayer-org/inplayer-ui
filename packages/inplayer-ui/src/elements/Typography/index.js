// @flow
import React, { type Node } from 'react';
import styled, { css } from 'styled-components';
import { prop, ifProp } from 'styled-tools';
import { uiColors, fontWeights, fontSizes } from 'utils';
import { textPrimary, textDanger, textSuccess, textWarning } from 'modifiers';
import { applyStyleModifiers } from 'styled-components-modifiers';

const typographyModifiers = {
  textPrimary,
  textDanger,
  textSuccess,
  textWarning,
};

type TypographyVariant = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';

const Paragraph = styled.p`
  font-size: ${fontSizes('medium')};
  color: ${ifProp('color', prop('color'), uiColors('text.main'))};
  ${ifProp(
    'description',
    css`
      color: ${uiColors('text.light')};
      font-size: ${fontSizes('large')};
    `
  )};
  ${applyStyleModifiers(typographyModifiers)};
`;

const Heading = styled.h1`
  font-family: ${prop('theme.font.primary')};
  font-weight: ${fontWeights('thin')};
  color: ${ifProp('color', prop('color'), uiColors('secondary.main'))};
  font-size: ${({ theme, variant }) => theme.font.sizes[variant]};
  ${applyStyleModifiers(typographyModifiers)};
`;

export type TypographyProps = {
  variant: TypographyVariant,
  description?: boolean,
  children: Node,
  /** A className can be passed down for further styling or extending with CSS-in-JS */
  className?: string,
  modifiers?: Array<String>,
  color?: string,
  style?: Object,
};

const Typography = ({
  variant,
  children,
  description,
  className,
  modifiers,
  color,
  style,
}: TypographyProps) => {
  if (variant === 'p') {
    return (
      <Paragraph
        className={className}
        description={description}
        modifiers={modifiers}
        color={color}
        style={style}
      >
        {children}
      </Paragraph>
    );
  }
  return (
    <Heading
      as={variant}
      className={className}
      variant={variant}
      modifiers={modifiers}
      color={color}
      style={style}
    >
      {children}
    </Heading>
  );
};

Typography.defaultProps = {
  className: '',
  modifiers: [],
  color: '',
  style: {},
  description: false,
};

export default Typography;
