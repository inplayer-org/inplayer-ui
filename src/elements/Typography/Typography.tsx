import React, { ReactNode } from 'react';
import styled from 'styled-components';
import { fontWeights, fontSizes } from 'utils';
import colors from 'config/colors';
import { textPrimary, textDanger, textSuccess, textWarning } from 'modifiers';
import { applyStyleModifiers } from 'styled-components-modifiers';

const typographyModifiers = {
  textPrimary,
  textDanger,
  textSuccess,
  textWarning,
};

type TypographyVariant = 'h1' | 'h2' | 'h3' | 'h4' | 'h5' | 'h6' | 'p';

const defaultProps = {
  className: '',
  modifiers: [],
  color: '',
  style: {},
  description: false,
};

const Paragraph = styled.p<{ description?: any; modifiers?: any }>`
  color: ${({ description }) => (description ? colors.fontGray : colors.fontDarkGray)};
  font-size: ${({ description }) => (description ? fontSizes('large') : fontSizes('medium'))};
`;

const Heading = styled.h1<{
  variant: TypographyVariant;
  color?: string;
  modifiers?: Array<string>;
  description?: any;
}>`
  font-family: ${({ theme }) => theme.font.primary};
  font-weight: ${fontWeights('thin')};
  color: ${colors.blue};
  font-size: ${({ variant, theme }) => theme.font.sizes[variant]};
  ${applyStyleModifiers(typographyModifiers)};
`;

interface TypographyProps {
  variant: TypographyVariant;
  description?: boolean;
  children: ReactNode;
  /** A className can be passed down for further styling or extending with CSS-in-JS */
  className?: string;
  modifiers?: Array<string>;
  color?: string;
  style?: Record<string, any>;
}

const Typography = ({
  variant,
  children,
  description,
  className,
  modifiers,
  color,
  style,
}: TypographyProps) => {
  if (variant === 'p') {
    return (
      <Paragraph
        className={className}
        description={description}
        modifiers={modifiers}
        color={color}
        style={style}
      >
        {children}
      </Paragraph>
    );
  }
  return (
    <Heading
      as={variant}
      className={className}
      variant={variant}
      modifiers={modifiers}
      color={color}
      style={style}
    >
      {children}
    </Heading>
  );
};

Typography.defaultProps = defaultProps;

export default Typography;
