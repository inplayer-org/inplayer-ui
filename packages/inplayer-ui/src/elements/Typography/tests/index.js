import { generateSnapshots } from 'packages/inplayer-ui/src/elements/Typography/tests/helpers';

import Typography from '../index';

generateSnapshots(Typography, { variant: 'h1' });
