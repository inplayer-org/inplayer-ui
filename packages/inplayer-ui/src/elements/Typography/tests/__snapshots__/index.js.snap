// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography renders correctly 1`] = `
.c0 {
  font-family: Roboto;
  font-weight: 100;
  color: #286090;
  font-size: 2.5rem;
}

<h1
  className="c0"
  color=""
  style={Object {}}
/>
`;
